core/schema/module.go

```diff
@@ -1152,13 +1152,35 @@ func (s *moduleSchema) moduleSource(ctx context.Context, query *core.Query, args
 			if args.Stable {
 				return nil, fmt.Errorf("no version provided for stable remote ref: %s", args.RefString)
 			}
-
 			var err error
 			modVersion, err = defaultBranch(ctx, cloneURL)
 			if err != nil {
 				return nil, fmt.Errorf("determine default branch: %w", err)
 			}
 		}
+		src.AsGitSource.Value.Version = modVersion
+
+		var subPath string
+		if len(segments) == 4 {
+			subPath = segments[3]
+		} else {
+			subPath = "/"
+		}
+		src.AsGitSource.Value.Subpath = subPath
+
+		commitRef := modVersion
+		if hasVersion && isSemver(modVersion) {
+			allTags, err := gitTags(ctx, cloneURL)
+			if err != nil {
+				return nil, fmt.Errorf("get git tags: %w", err)
+			}
+			matched, err := matchVersion(allTags, modVersion, subPath)
+			if err != nil {
+				return nil, fmt.Errorf("matching version to tags: %w", err)
+			}
+			// reassign modVersion to matched tag which could be subPath/tag
+			commitRef = matched
```

Comment:
Just double checking that we want to leave `src.AsGitSource.Value.Version` alone in this case. Basically, even though we resolved e.g. `v1.0.0` to `path/v1.0.0`, we still leave `Version` as what the user provided.

Very grey area for me, not clear one or the other is preferable, mostly comes down to daggerverse/analytics needs, so just want to call out in case significant.

Reply:
Good question - yeah, I get the same sense about all of this being filled with grey areas. I changed my mind a few times while writing this comment; hopefully I've navigated the grey somewhat and added some guideposts.

From the perspective of Daggerverse + Analytics I think the most valuable thing here is actually a precise semver version. I originally imagined it as 'preserving the original user input' as the field currently documents. That will incidentally be true since the user originally inputs just `v1.2.3` and not `path/v1.2.3` as the version, so I think the change is good as-is.

Justification:

* In Daggerverse we're only interested in displaying the logical semantic version; the path prefixes are just an implementation detail of how Git sources interpret semver, so `asString` shouldn't show them, and neither should the Daggerverse UI.

* In analytics we want to record what version of a module was used, and it seems like we would always want an exact version there. Someday we want to support semver shorthand like `v1` => `v1.2.3`, and if we just kept this value as the original user input, we'd record separate `v1`, `v1.2`, and `v1.2.3` metrics which don't help you track module version usage. (It _does_ let you track how precise people are being with their version selection, but that feels like a lower priority metric atm.)

But, I'm also sympathetic to keeping all the intermediate forms in the surprisingly complex task of `ref` resolution around to support integrity checks or further troubleshooting.

So, someday, if you resolve `github.com/vito/daggerverse/apko:v1` or `vito/apko@v1` and the latest version is `v1.2.3`, maybe you could get this back:

* ModuleSource.Version = `v1.2.3`
* ModuleSource.RequestedVersion = `v1`
* ModuleSource.AsGitSource.CloneURL = `https://github.com/vito/daggerverse`
* ModuleSource.AsGitSource.Commit = `abcdef`
* ModuleSource.AsGitSource.Ref = `apko/v1.2.3`
* ModuleSource.AsGitSource.Version = `v1.2.3`
* ModuleSource.Subpath = `/apko`

...and we'd put some subset that into your `dagger.json` in raw object form so we don't have to worry about getting the ref syntax right the first time. :sweat_smile: 

Lemme know if I'm missing something.

Reply:
> So, someday, if you resolve github.com/vito/daggerverse/apko:v1 or vito/apko@v1 and the latest version is v1.2.3, maybe you could get this back:

Yep all of the above makes sense to me and yeah in the longer term we can easily update this to track every little detail as needed 👍